mlflow/projects/__init__.py

```diff
@@ -145,7 +147,34 @@ def _run(uri, experiment_id, entry_point="main", version=None, parameters=None,
             work_dir=work_dir, entry_point=entry_point, parameters=parameters,
             experiment_id=experiment_id,
             use_conda=use_conda, storage_dir=storage_dir, run_id=active_run.info.run_id)
-    supported_backends = ["local", "databricks"]
+    elif backend == "kubernetes":
+        from mlflow.projects import kubernetes as kb
+        tracking.MlflowClient().set_tag(active_run.info.run_uuid,
+                                        MLFLOW_PROJECT_ENV,
+                                        "kubernetes")
+        _validate_docker_installation()
+        if not project.docker_env.get('image'):
+            raise ExecutionException("Project with docker environment must specify the docker "
+                                     "image to use via an 'image' field under the 'docker_env' "
+                                     " field.")
+        kube_config = _parse_kubernetes_config(backend_config, work_dir)
+        image = _build_docker_image(work_dir=work_dir,
+                                    image_uri=kube_config["image-uri"],
+                                    base_image=project.docker_env.get('image'),
+                                    active_run=active_run)
+        kb.push_image_to_registry(image)
+        job_info = kb.run_kubernetes_job(image,
+                                         _get_entry_point_command(project, entry_point,
+                                                                  parameters, storage_dir),
+                                         _get_run_env_vars(
+                                             run_id=active_run.info.run_uuid,
+                                             experiment_id=active_run.info.experiment_id),
+                                         kube_config['kube-context'],
+                                         kube_config['kube-job-template'])
+        return kb.monitor_job_status(job_info["job_name"],
```

Comment:
I think we need to return a `SubmittedRun` object here which implements a wait function which actually implements the polling for the completion of the k8s project.

For example: https://github.com/mlflow/mlflow/blob/master/mlflow/projects/databricks.py#L325.

Reply:
I see, but the point here is that the monitor is reading a stream from pod's log. It's not polling an REST API as we do in Databricks mode. We can think about move this logic to a `SubmittedRun` object though.

What do you think?

Reply:
It's similar, when the function `run` invokes `_run`, it expects for `_run` to return a `SubmittedRun` object which it can wait/poll for completion on. The nice thing about this interface is now we can run mlflow jobs asynchronously by simply not blocking for the `SubmittedRun`'s completion.